plugin.go

```diff
@@ -28,40 +45,259 @@ func (p *Plugin) PluginInfo() launchr.PluginInfo {
 
 // OnAppInit implements launchr.Plugin interface.
 func (p *Plugin) OnAppInit(app launchr.App) error {
+	app.GetService(&p.cfg)
+
 	p.app = app
 	return nil
 }
 
 // CobraAddCommands implements launchr.CobraPlugin interface to provide web functionality.
 func (p *Plugin) CobraAddCommands(rootCmd *cobra.Command) error {
 	// Flag options.
-	var port string
+	var port int
 	var proxyClient string
 	var useSwaggerUI bool
+	var foreground bool
 	var cmd = &cobra.Command{
-		Use:   "web",
-		Short: "Starts web server",
+		Use:       "web",
+		Short:     "Starts web server",
+		Args:      cobra.MatchAll(cobra.RangeArgs(0, 1), cobra.OnlyValidArgs),
+		ValidArgs: []string{stopArg},
 		RunE: func(cmd *cobra.Command, args []string) error {
 			// Don't show usage help on a runtime error.
 			cmd.SilenceUsage = true
 
+			pluginTmpDir := p.getPluginTempDir()
+			pidFile := fmt.Sprintf("%s/%s.pid", pluginTmpDir, strings.ToLower(cmd.Name()))
+
+			if !isBackGroundEnv() {
+				runInfo, err := server.GetRunInfo(pluginTmpDir)
+				if err != nil {
+					return err
+				}
+
+				// If 'stop' arg passed, try to interrupt process and remove PID file.
+				if len(args) > 0 && args[0] == stopArg {
+					err = stopWeb(pidFile, runInfo)
+					if err != nil {
+						return err
+					}
+
+					cli.Println("Web server stopped successfully")
+					return nil
+				}
+
+				if isWebRunning(pidFile, runInfo) {
+					return fmt.Errorf("server is already running by address %s, please stop it before starting new one", runInfo.BaseURL)
+				}
+			} else {
+				// Store background outputs into plugin tmp folder.
+				err := redirectOutputs(pluginTmpDir)
+				if err != nil {
+					return err
+				}
+			}
+
+			if !server.IsPortFree(port) {
+				log.Info("web server port %d you are trying to use is not available", port)
+				port = server.GetFreePort(port, 20)
+			}
+
 			runOpts := &server.RunOptions{
-				Addr:        fmt.Sprintf(":%s", port), // @todo use proper addr
-				APIPrefix:   APIPrefix,
-				SwaggerJSON: useSwaggerUI,
-				ProxyClient: proxyClient,
+				Addr:          fmt.Sprintf(":%d", port), // @todo use proper addr
+				APIPrefix:     APIPrefix,
+				SwaggerJSON:   useSwaggerUI,
+				ProxyClient:   proxyClient,
+				RunInfoFolder: pluginTmpDir,
 				// @todo use embed fs for client or provide path ?
 			}
-
 			prepareRunOption(p, runOpts)
 
+			if foreground {
+				return server.Run(cmd.Context(), p.app, runOpts)
+			}
+
+			if !isBackGroundEnv() {
+				pid, err := p.runAsBackgroundCmd(cmd, pidFile)
+				if err != nil {
+					return err
+				}
+				// Wait until background server is up.
+				// Check if run info created and server is reachable.
+				// Print server URL in CLI.
+				// Kill process in case of timeout
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+
+				var wg sync.WaitGroup
+				wg.Add(1)
+				go func() {
```

Comment:
It's an overkill
1. You don't need WaitGroup. WaitGroup is for cases when you try to join multiple threads, you have only 1
2. You don't need to make a go function. You are spawning a goroutine and then block with WG. You can just do for with `break`
3. Your logic is strange. First of all, you don't need context, because context is intended to be passed to nested functions to control them. You don't change context of a function that's why it's better to control directly with a timer. 
5. You run a timer with context and you run a ticker that checks every 5 secs, which is misleading. Check tickers https://gobyexample.com/tickers . Maybe I would still not use goroutines, and only `for` loop with 2 timers global 5 sec and a ticker. Global goes first, because `select` picks first available in order like switch.

I think the whole implementation is strange.
We have the following possibilities here:
1. Use bash & functionality. If bash session stops, all processes are killed and not become unmanaged. I like this solution the most.
2. Use screen/tmux if you need to save the session
2. Use https://github.com/sevlyar/go-daemon . It has all your code but better supports all OSes for daemonising a proccess

I already investigated such functionality for dev. I think using daemon must be a last resort.
I insist on using bash bg or tmux. They require some skills from users, but are much better than a headache of supporting daemon functionality which may be buggy in implementation.
Such things must be managed by OS with systemctl, not application.

We must understand, that generating each web ui for each action - is NOT OK. And such functionality must be implemented inside launchr, not by generating unmanaged processes.

Reply:
> (...) using bash bg or tmux. They require some skills from users, but (...)

Main thing is that the audience of this functionality is precisely low-skilled users: clients, funcitonal teams and so on. Soooo if we don't make it easy for them to launch/use launchr ui, they won't us it at all.

Tech-savy users will mostly use the cli interface, they can choose to use the wbe ui but it's not meant for them